From 6623991e7f7f14678e39bf2ed8d16ed36fbe07d4 Mon Sep 17 00:00:00 2001
From: "Roy.Li" <rongqing.li@windriver.com>
Date: Wed, 1 Jun 2016 11:12:48 -0400
Subject: watchdog: Fix ping mode failure

Upstream-Status: Pending

When watchdog works on ping mode, the system will be rebooted since
watchdog can not receive the expected ECOREPLY on a setting interval.

Ping mode uses a raw socket to send a ECO packet, then uses select()
to wait and recvfrom() to receive the ECOREPLY packet, if select()
shows the data is ready, and the data is not the expected ECOREPLY,
and waiting time is not overdue, it will continue use select() and
recvfrom().

Problem is that the raw socket can receive any icmp packets, if we do
not set filters, and there are many icmp packets on socket, this
program will not find its interested ECOREPLY packet in a special
interval, which makes the ping mode fail.

Other program is that watchdog sometime can not reach the call of
recvfrom to try to receive packets since tv_sec of struct timeval
of select parameter is 0.

The timeout of select() is the result of ping interval minusing the
time of calling gettimeofday spending, when ping interval is 1 second,
and the call of gettimeofday() spends several useconds, the tv_sec of
struct timeval of select parameter must be 0, at that condition, we
should it is valid of tv_sec of struct timeval of select parameter be 0

Signed-off-by: Roy.Li <rongqing.li@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>
---
 src/watchdog.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/src/watchdog.c b/src/watchdog.c
index bab7f82..e24c947 100644
--- a/src/watchdog.c
+++ b/src/watchdog.c
@@ -28,6 +28,7 @@
 #include <sys/types.h>
 #include <sys/ioctl.h>
 #include <linux/oom.h>
+#include <linux/icmp.h>
 #include <linux/watchdog.h>
 #include <string.h>
 
@@ -573,6 +574,8 @@ int main(int argc, char *const argv[])
     pid_t child_pid;
     int oom_adjusted = 0;
     struct stat s;
+    struct icmp_filter filt;
+    filt.data = ~(1<<ICMP_ECHOREPLY);
 
 #if USE_SYSLOG
     char *opts = "d:i:n:Ffsvbql:p:t:c:r:m:a:";
@@ -709,7 +712,7 @@ int main(int argc, char *const argv[])
 		perror(progname);
 		exit(1);
 	    }
-	    
+	    setsockopt(net->sock_fp, SOL_RAW, ICMP_FILTER, (char*)&filt, sizeof(filt));
 	    /* this is necessary for broadcast pings to work */
 	    (void) setsockopt(net->sock_fp, SOL_SOCKET, SO_BROADCAST, (char *)&hold, sizeof(hold));
     
-- 
2.7.4

